File: frontend/src/features/deploy/api/deploy.ts
import { http } from '@/shared/http/http'
import type { ApiResponse, PageResult } from '@/shared/types/api'

async function unwrap<T>(request: Promise<unknown>): Promise<T> {
  const resp = (await request) as ApiResponse<T>
  return resp.data
}

export interface DeployServerItem {
  id: number
  name: string
  ip: string
  ssh_port: number
  user: string
  auth_type: 'password' | 'key'
  credential_id?: number
  os?: string
  os_version?: string
  kernel?: string
  cpu_cores?: number
  memory_mb?: number
  disk_gb?: number
  status: string
  remark?: string
  created_at: string
  updated_at: string
}

export interface SSHCredentialItem {
  id: number
  name: string
  auth_type: 'password' | 'key'
  username: string
  remark?: string
  server_count: number
  created_at: string
  updated_at: string
}

export interface DeployPlanNodeItem {
  id: number
  server_id: number
  role: 'master' | 'worker'
  sort_order: number
}

export interface DeployPlanItem {
  id: number
  name: string
  cluster_name: string
  k8s_version: string
  pod_cidr: string
  svc_cidr: string
  cni_type: 'flannel' | 'calico' | 'cilium'
  status: string
  task_id?: number
  cluster_id?: number
  nodes?: DeployPlanNodeItem[]
  created_at: string
  updated_at: string
}

export interface CreateDeployServerReq {
  name: string
  ip: string
  ssh_port: number
  user: string
  auth_type: 'password' | 'key'
  credential_id?: number
  credential: string
  remark?: string
}

export interface CreateSSHCredentialReq {
  name: string
  auth_type: 'password' | 'key'
  username: string
  credential: string
  remark?: string
}

export interface CreateDeployPlanReq {
  name: string
  cluster_name: string
  k8s_version: string
  pod_cidr: string
  svc_cidr: string
  cni_type: 'flannel' | 'calico' | 'cilium'
  nodes: Array<{
    server_id: number
    role: 'master' | 'worker'
    sort_order: number
  }>
}

export function listDeployServers(params: { page?: number; page_size?: number; keyword?: string; status?: string } = {}) {
  return unwrap<PageResult<DeployServerItem>>(http.get('/api/v1/deploy/servers', { params }))
}

export function createDeployServer(data: CreateDeployServerReq) {
  return unwrap<{ id: number }>(http.post('/api/v1/deploy/servers', data))
}

export function updateDeployServer(id: number, data: Partial<CreateDeployServerReq>) {
  return unwrap<null>(http.put(`/api/v1/deploy/servers/${id}`, data))
}

export function deleteDeployServer(id: number) {
  return unwrap<null>(http.delete(`/api/v1/deploy/servers/${id}`))
}

export interface SSHProbeResult {
  status: string
  message: string
  os?: string
  os_version?: string
  kernel?: string
}

export function testDeployServerSSH(id: number) {
  return unwrap<SSHProbeResult>(http.post(`/api/v1/deploy/servers/${id}/test-ssh`))
}

export function listSSHCredentials(params: { page?: number; page_size?: number; keyword?: string; auth_type?: string } = {}) {
  return unwrap<PageResult<SSHCredentialItem>>(http.get('/api/v1/deploy/credentials', { params }))
}

export function createSSHCredential(data: CreateSSHCredentialReq) {
  return unwrap<{ id: number }>(http.post('/api/v1/deploy/credentials', data))
}

export function getSSHCredential(id: number) {
  return unwrap<SSHCredentialItem>(http.get(`/api/v1/deploy/credentials/${id}`))
}

export function updateSSHCredential(id: number, data: Partial<CreateSSHCredentialReq>) {
  return unwrap<null>(http.put(`/api/v1/deploy/credentials/${id}`, data))
}

export function deleteSSHCredential(id: number) {
  return unwrap<null>(http.delete(`/api/v1/deploy/credentials/${id}`))
}

export function batchDeleteSSHCredentials(ids: number[]) {
  return unwrap<null>(http.post('/api/v1/deploy/credentials/batch-delete', { ids }))
}

export function listDeployPlans(params: { page?: number; page_size?: number; keyword?: string; status?: string } = {}) {
  return unwrap<PageResult<DeployPlanItem>>(http.get('/api/v1/deploy/plans', { params }))
}

export function createDeployPlan(data: CreateDeployPlanReq) {
  return unwrap<{ id: number }>(http.post('/api/v1/deploy/plans', data))
}

export function deleteDeployPlan(id: number) {
  return unwrap<null>(http.delete(`/api/v1/deploy/plans/${id}`))
}

export function executeDeployPlan(id: number) {
  return unwrap<{ task_id: number }>(http.post(`/api/v1/deploy/plans/${id}/execute`))
}

export interface DryRunStep {
  key: string
  title: string
  description: string
  phase: string
  commands: string[]
  depends_on?: string[]
}
export interface DryRunNodeFlow {
  server_id: number
  server_name: string
  ip: string
  role: string
  steps: DryRunStep[]
}
export interface DryRunResult {
  plan_id: number
  plan_name: string
  cluster_name: string
  k8s_version: string
  cni_type: string
  nodes: DryRunNodeFlow[]
  summary: Record<string, number>
}
export function dryRunDeployPlan(id: number) {
  return unwrap<DryRunResult>(http.get(`/api/v1/deploy/plans/${id}/dry-run`))
}

export function cancelDeployPlan(id: number) {
  return unwrap<null>(http.post(`/api/v1/deploy/plans/${id}/cancel`))
}

export function retryDeployPlan(id: number) {
  return unwrap<{ task_id: number }>(http.post(`/api/v1/deploy/plans/${id}/retry`))
}

// SSE 实时日志订阅
export function subscribeDeployLogs(taskId: number, onLog: (log: string) => void, onDone?: (status: string, message: string) => void): EventSource {
  const token = localStorage.getItem('token') || ''
  const url = `/api/v1/deploy/tasks/${taskId}/logs/sse`
  
  const eventSource = new EventSource(url)
  
  eventSource.onmessage = (event) => {
    try {
      const data = JSON.parse(event.data)
      if (data.log) {
        onLog(data.log)
      }
    } catch (e) {
      console.error('Failed to parse SSE log:', e)
    }
  }
  
  eventSource.addEventListener('done', (event) => {
    try {
      const data = JSON.parse(event.data)
      onDone?.(data.status, data.message)
    } catch (e) {
      console.error('Failed to parse SSE done event:', e)
    }
    eventSource.close()
  })
  
  eventSource.onerror = (error) => {
    console.error('SSE connection error:', error)
    eventSource.close()
  }
  
  return eventSource
}

export interface TaskLogsResult {
  task_id: number
  logs: string[]
  total: number
}

export function getDeployTaskLogs(taskId: number, params?: { offset?: number; limit?: number }) {
  return unwrap<TaskLogsResult>(http.get(`/api/v1/deploy/tasks/${taskId}/logs`, { params }))
}

// ─── 部署配置管理 ───

export interface DeployConfigItem {
  id: number
  step_key: string
  os_type: string
  step_name: string
  step_order: number
  command_template: string
  description?: string
  enabled: boolean
  timeout_seconds: number
  retry_count: number
  created_by: number
  created_at: string
  updated_at: string
}

export interface DeployConfigVersion {
  id: number
  config_id: number
  step_key: string
  command_template: string
  description?: string
  change_type: string
  changed_by: number
  changed_at: string
  change_summary?: string
}

export function listDeployConfigs(params?: { os_type?: string }) {
  return unwrap<DeployConfigItem[]>(http.get('/api/v1/deploy/configs', { params }))
}

export function listSupportedOSTypes() {
  return unwrap<string[]>(http.get('/api/v1/deploy/configs/os-types'))
}

export function getDeployConfig(id: number) {
  return unwrap<DeployConfigItem>(http.get(`/api/v1/deploy/configs/${id}`))
}

export function updateDeployConfig(id: number, data: {
  command_template: string
  description?: string
  enabled?: boolean
  timeout_seconds?: number
  retry_count?: number
  change_summary?: string
}) {
  return unwrap<null>(http.put(`/api/v1/deploy/configs/${id}`, data))
}

export function getDeployConfigVersions(id: number) {
  return unwrap<DeployConfigVersion[]>(http.get(`/api/v1/deploy/configs/${id}/versions`))
}

// ─── 仓库配置管理 ───

export interface RepositoryItem {
  id: number
  repo_type: string
  name: string
  url: string
  description?: string
  auth_type: string
  priority: number
  enabled: boolean
  is_default: boolean
  mirror_of?: string
  created_by: number
  created_at: string
  updated_at: string
}

export function listRepositories(params?: { type?: string }) {
  return unwrap<RepositoryItem[]>(http.get('/api/v1/deploy/repositories', { params }))
}

export function getRepository(id: number) {
  return unwrap<RepositoryItem>(http.get(`/api/v1/deploy/repositories/${id}`))
}

export function createRepository(data: {
  repo_type: string
  name: string
  url: string
  description?: string
  auth_type?: string
  priority?: number
  is_default?: boolean
  mirror_of?: string
}) {
  return unwrap<{ id: number }>(http.post('/api/v1/deploy/repositories', data))
}

export function updateRepository(id: number, data: {
  name?: string
  url?: string
  description?: string
  auth_type?: string
  priority?: number
  enabled?: boolean
  is_default?: boolean
  mirror_of?: string
}) {
  return unwrap<null>(http.put(`/api/v1/deploy/repositories/${id}`, data))
}

export function deleteRepository(id: number) {
  return unwrap<null>(http.delete(`/api/v1/deploy/repositories/${id}`))
}

File: backend/internal/model/deploy.go
package model

import "time"

type DeployServer struct {
	ID            uint64     `gorm:"column:id;primaryKey;autoIncrement"`
	Name          string     `gorm:"column:name"`
	IP            string     `gorm:"column:ip"`
	SSHPort       int        `gorm:"column:ssh_port"`
	User          string     `gorm:"column:user"`
	AuthType      string     `gorm:"column:auth_type"`
	CredentialID  *uint64    `gorm:"column:credential_id"`
	CredentialEnc string     `gorm:"column:credential_enc"`
	OS            *string    `gorm:"column:os"`
	OSVersion     *string    `gorm:"column:os_version"`
	Kernel        *string    `gorm:"column:kernel"`
	CPUCores      *uint      `gorm:"column:cpu_cores"`
	MemoryMB      *uint64    `gorm:"column:memory_mb"`
	DiskGB        *uint64    `gorm:"column:disk_gb"`
	Status        string     `gorm:"column:status"`
	Labels        JSONMap    `gorm:"column:labels;type:json"`
	Remark        *string    `gorm:"column:remark"`
	CreatedAt     time.Time  `gorm:"column:created_at;autoCreateTime"`
	UpdatedAt     time.Time  `gorm:"column:updated_at;autoUpdateTime"`
	DeletedAt     *time.Time `gorm:"column:deleted_at"`
}

func (DeployServer) TableName() string { return "deploy_servers" }

type SSHCredential struct {
	ID            uint64     `gorm:"column:id;primaryKey;autoIncrement"`
	Name          string     `gorm:"column:name"`
	AuthType      string     `gorm:"column:auth_type"`
	Username      string     `gorm:"column:username"`
	CredentialEnc string     `gorm:"column:credential_enc"`
	Remark        *string    `gorm:"column:remark"`
	CreatedAt     time.Time  `gorm:"column:created_at;autoCreateTime"`
	UpdatedAt     time.Time  `gorm:"column:updated_at;autoUpdateTime"`
	DeletedAt     *time.Time `gorm:"column:deleted_at"`
}

func (SSHCredential) TableName() string { return "ssh_credentials" }

type DeployPlan struct {
	ID               uint64          `gorm:"column:id;primaryKey;autoIncrement"`
	Name             string          `gorm:"column:name"`
	ClusterName      string          `gorm:"column:cluster_name"`
	K8sVersion       string          `gorm:"column:k8s_version"`
	PodCIDR          string          `gorm:"column:pod_cidr"`
	SvcCIDR          string          `gorm:"column:svc_cidr"`
	CNIType          string          `gorm:"column:cni_type"`
	CNIConfig        JSONMap         `gorm:"column:cni_config;type:json"`
	Addons           JSONStringSlice `gorm:"column:addons;type:json"`
	StepOverrides    JSONMap         `gorm:"column:step_overrides;type:json"`
	PreflightIgnores JSONStringSlice `gorm:"column:preflight_ignores;type:json"`
	Status           string          `gorm:"column:status"`
	TaskID           *uint64         `gorm:"column:task_id"`
	ClusterID        *uint64         `gorm:"column:cluster_id"`
	CreatedBy        uint64          `gorm:"column:created_by"`
	CreatedAt        time.Time       `gorm:"column:created_at;autoCreateTime"`
	UpdatedAt        time.Time       `gorm:"column:updated_at;autoUpdateTime"`
	DeletedAt        *time.Time      `gorm:"column:deleted_at"`
}

func (DeployPlan) TableName() string { return "deploy_plans" }

type DeployPlanNode struct {
	ID        uint64    `gorm:"column:id;primaryKey;autoIncrement"`
	PlanID    uint64    `gorm:"column:plan_id"`
	ServerID  uint64    `gorm:"column:server_id"`
	Role      string    `gorm:"column:role"`
	SortOrder int       `gorm:"column:sort_order"`
	CreatedAt time.Time `gorm:"column:created_at;autoCreateTime"`
}

func (DeployPlanNode) TableName() string { return "deploy_plan_nodes" }

type DeployPlanStepOverride struct {
	StepKey         string  `json:"step_key"`
	NodeRole        string  `json:"node_role,omitempty"`
	NodeServerID    *uint64 `json:"node_server_id,omitempty"`
	CommandTemplate string  `json:"command_template"`
	Description     *string `json:"description,omitempty"`
	TimeoutSeconds  *int    `json:"timeout_seconds,omitempty"`
	RetryCount      *int    `json:"retry_count,omitempty"`
	Enabled         *bool   `json:"enabled,omitempty"`
}

type DeployLog struct {
	ID        uint64    `gorm:"column:id;primaryKey;autoIncrement"`
	PlanID    uint64    `gorm:"column:plan_id"`
	TaskID    *uint64   `gorm:"column:task_id"`
	StepName  *string   `gorm:"column:step_name"`
	ServerIP  *string   `gorm:"column:server_ip"`
	Level     string    `gorm:"column:level"`
	Message   string    `gorm:"column:message"`
	CreatedAt time.Time `gorm:"column:created_at;autoCreateTime"`
}

func (DeployLog) TableName() string { return "deploy_logs" }

// DeployConfig 部署流程配置模型
type DeployConfig struct {
	ID              uint64    `gorm:"column:id;primaryKey;autoIncrement" json:"id"`
	StepKey         string    `gorm:"column:step_key" json:"step_key"`
	OSType          string    `gorm:"column:os_type;default:ubuntu" json:"os_type"`
	StepName        string    `gorm:"column:step_name" json:"step_name"`
	StepOrder       int       `gorm:"column:step_order" json:"step_order"`
	CommandTemplate string    `gorm:"column:command_template;type:text" json:"command_template"`
	Description     *string   `gorm:"column:description;type:text" json:"description"`
	Enabled         bool      `gorm:"column:enabled;default:true" json:"enabled"`
	TimeoutSeconds  int       `gorm:"column:timeout_seconds;default:600" json:"timeout_seconds"`
	RetryCount      int       `gorm:"column:retry_count;default:0" json:"retry_count"`
	CreatedBy       uint64    `gorm:"column:created_by" json:"created_by"`
	CreatedAt       time.Time `gorm:"column:created_at;autoCreateTime" json:"created_at"`
	UpdatedAt       time.Time `gorm:"column:updated_at;autoUpdateTime" json:"updated_at"`
}

func (DeployConfig) TableName() string { return "deploy_configs" }

// DeployConfigVersion 部署配置版本历史模型
type DeployConfigVersion struct {
	ID              uint64    `gorm:"column:id;primaryKey;autoIncrement" json:"id"`
	ConfigID        uint64    `gorm:"column:config_id" json:"config_id"`
	StepKey         string    `gorm:"column:step_key" json:"step_key"`
	CommandTemplate string    `gorm:"column:command_template;type:text" json:"command_template"`
	Description     *string   `gorm:"column:description;type:text" json:"description"`
	ChangeType      string    `gorm:"column:change_type" json:"change_type"`
	ChangedBy       uint64    `gorm:"column:changed_by" json:"changed_by"`
	ChangedAt       time.Time `gorm:"column:changed_at;autoCreateTime" json:"changed_at"`
	ChangeSummary   *string   `gorm:"column:change_summary" json:"change_summary"`
}

func (DeployConfigVersion) TableName() string { return "deploy_config_versions" }

// DeployRepository 仓库配置模型
type DeployRepository struct {
	ID          uint64     `gorm:"column:id;primaryKey;autoIncrement" json:"id"`
	RepoType    string     `gorm:"column:repo_type" json:"repo_type"`
	Name        string     `gorm:"column:name" json:"name"`
	URL         string     `gorm:"column:url" json:"url"`
	Description *string    `gorm:"column:description;type:text" json:"description"`
	AuthType    string     `gorm:"column:auth_type;default:none" json:"auth_type"`
	AuthConfig  JSONMap    `gorm:"column:auth_config;type:json" json:"auth_config"`
	Priority    int        `gorm:"column:priority;default:100" json:"priority"`
	Enabled     bool       `gorm:"column:enabled;default:true" json:"enabled"`
	IsDefault   bool       `gorm:"column:is_default;default:false" json:"is_default"`
	MirrorOf    *string    `gorm:"column:mirror_of" json:"mirror_of"`
	CreatedBy   uint64     `gorm:"column:created_by" json:"created_by"`
	CreatedAt   time.Time  `gorm:"column:created_at;autoCreateTime" json:"created_at"`
	UpdatedAt   time.Time  `gorm:"column:updated_at;autoUpdateTime" json:"updated_at"`
	DeletedAt   *time.Time `gorm:"column:deleted_at" json:"deleted_at,omitempty"`
}

func (DeployRepository) TableName() string { return "deploy_repositories" }
